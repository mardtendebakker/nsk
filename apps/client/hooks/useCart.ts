import { useEffect, useState } from 'react';
import cartStore, {
  ADD_MODULE_REQUEST,
  CLEAR_CART_REQUEST,
  Cart, EVENTS, Module, REMOVE_MODULE_REQUEST,
} from '../stores/cart';

const useCart = (): {
  state: Cart,
  removeModule:(module: Module)=> void,
  addModule:(module: Module)=> void,
  clearCart:()=> void,
  totalModulesCount: () => number
  totalAmount: () => number
} => {
  const [state, setState] = useState<Cart>(cartStore.cart);

  function cb() {
    setState(cartStore.cart);
  }

  useEffect(() => {
    EVENTS.forEach((event) => {
      cartStore.on(event, cb);
    });

    return () => {
      EVENTS.forEach((event) => {
        cartStore.removeListener(event, cb);
      });
    };
  }, []);

  return {
    state,
    removeModule: (module: Module) => {
      cartStore.emit(REMOVE_MODULE_REQUEST, module);
    },
    addModule: (module: Module) => {
      cartStore.emit(ADD_MODULE_REQUEST, module);
    },
    clearCart: () => {
      cartStore.emit(CLEAR_CART_REQUEST);
    },
    totalModulesCount: () => state.modules.length,
    totalAmount: () => {
      let total = 0;

      state.modules.forEach(({ price }) => {
        total += price;
      });

      return total;
    },
  };
};

export default useCart;
