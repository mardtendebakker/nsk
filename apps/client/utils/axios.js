import axios from 'axios';

export { AxiosPromise, AxiosResponse, AxiosError } from 'axios';

const baseURL = process.env.NODE_ENV === 'development' ? 'http://localhost:3333' : 'production_url';

export const client = axios.create({
  baseURL,
  timeout: 120000,
  headers: {
    Accept: 'application/json',
    'Content-Type': 'application/json',
  },
});

export const { CancelToken } = axios;

export const SUPLIERS_PATH = '';

export default client;
