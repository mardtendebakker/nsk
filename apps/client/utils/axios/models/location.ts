export interface Location {
  id:number,
  name:string,
  zipcodes:string
  location_template: { id: number, template: string }[]
}
