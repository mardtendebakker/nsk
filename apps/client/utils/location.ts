export function getQueryParam(param: string, defaultValue?: string): null | undefined | string {
  return Object.fromEntries(
    // eslint-disable-next-line no-restricted-globals
    new URLSearchParams(location.search) as unknown as [[string, string]],
  )[param] || defaultValue || undefined;
}
