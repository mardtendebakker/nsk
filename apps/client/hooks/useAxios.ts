import { useEffect, useRef, useState } from 'react';
import { useSnackbar } from 'notistack';
import { trans } from 'itranslator';
import axios, {
  AxiosResponse, AxiosPromise, CancelToken, AxiosError, CanceledError,
} from '../utils/axios';
import { hideProgress, showProgress } from '../components/topLinearProgress';

export const GET = 'get';
export const POST = 'post';
export const PATCH = 'patch';
export const DELETE = 'delete';
export const PUT = 'put';

type Method = 'get' | 'post' | 'patch' | 'delete' | 'put';
type ResponseType = 'json' | 'blob';

const useAxios = <Data>(
  method: Method,
  path: string,
  {
    withProgressBar = false,
    showErrorMessage = true,
    showSuccessMessage = false,
    customSuccessMessage,
    defaultParams = {},
    customStatusesMessages,
  }
  : {
    withProgressBar? : boolean,
    showErrorMessage? : boolean,
    showSuccessMessage? : boolean,
    customSuccessMessage?: (response: AxiosResponse) => string
    defaultParams?: object,
    customStatusesMessages?: { [key: number]: string }
  } = {},
) : {
    data?: Data,
    response?: AxiosResponse,
    error?: AxiosError,
    performing: boolean,
    call: Call,
  } => {
  const [response, setResponse] = useState<AxiosResponse>();
  const [error, setError] = useState<AxiosError>();
  const [performing, setPerforming] = useState<boolean>(false);
  const { enqueueSnackbar } = useSnackbar();
  const source = useRef(CancelToken.source());

  useEffect(() => source.current.cancel, []);

  async function call(
    explicitPath: string,
    params?: object,
    body?: object,
    responseType?: ResponseType,
    headers?: object,
  )
    : AxiosPromise<AxiosResponse> {
    if (method === POST || method === PATCH || method === PUT) {
      return axios[method](explicitPath, body, {
        cancelToken: source.current.token,
        params: { ...defaultParams, ...params },
        responseType,
        headers,
      });
    }

    return axios[method](explicitPath, {
      cancelToken: source.current.token, params: { ...defaultParams, ...params }, data: body, responseType,
    });
  }

  function handleSuccess(handledResponse: AxiosResponse) {
    enqueueSnackbar(
      customSuccessMessage ? customSuccessMessage(handledResponse) : trans(handledResponse.status.toString()),
      { variant: 'success' },
    );
  }

  function handleError(e: Error | AxiosError) {
    const errorResponse = (e instanceof AxiosError && e.response) ? e.response : { status: 500, data: { message: undefined } };
    const errorMessage = customStatusesMessages && customStatusesMessages[errorResponse.status];
    let messages: string[] = [];

    switch (true) {
      case typeof errorMessage == 'string':
        messages.push(errorMessage);
        break;
      case Array.isArray(errorResponse.data.message):
        messages = errorResponse.data.message;
        break;
      case typeof errorResponse.data.message == 'string' && !!errorResponse.data.message:
        messages.push(errorResponse.data.message);
        break;
      default:
        messages.push(trans(errorResponse.status.toString()));
    }

    messages.forEach((message) => {
      enqueueSnackbar(message, { variant: 'error' });
    });
  }

  return {
    data: response?.data,
    response,
    error,
    performing,
    /** @throws {Error} */
    call: async (
      {
        params, body, path: explicitPath, responseType, headers,
      }
      : { params?: object, body?: object, path?: string, responseType?: ResponseType, headers?: object }
      = {},
      cb?: (e: Error, axiosResponse?: AxiosResponse) => void,
    ): Promise<AxiosResponse | void> => {
      try {
        const finalPath = explicitPath || path;

        setPerforming(true);
        if (withProgressBar) {
          showProgress();
        }

        const resp = await call(finalPath, params, body, responseType, headers);
        setResponse(resp);

        if (showSuccessMessage || customSuccessMessage) {
          handleSuccess(resp);
        }

        if (cb) {
          cb(null, resp);
        }

        return resp;
      } catch (e) {
        if (e instanceof CanceledError) {
          throw new Error(explicitPath + path);
        }

        setError(e);
        if (showErrorMessage) {
          handleError(e);
        }

        if (cb) {
          cb(e);
        } else {
          throw e;
        }
      } finally {
        setPerforming(false);

        if (withProgressBar) {
          hideProgress();
        }
      }

      return undefined;
    },
  };
};

export type Call = (
  arg0?: { params?: object, body?: object, path?: string, responseType?: ResponseType, headers?: object },
  cb?: (e: Error, axiosResponse?: AxiosResponse) => void,
) => Promise<AxiosResponse | void>;

export default useAxios;
