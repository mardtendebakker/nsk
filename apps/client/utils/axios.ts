import axios from 'axios';

export { AxiosError } from 'axios';
export type { AxiosPromise, AxiosResponse } from 'axios';

const baseURL = process.env.NODE_ENV === 'development' ? 'http://localhost:3333/api' : process.env.NX_AXIOS_BASE_URL;

export const client = axios.create({
  baseURL,
  timeout: 120000,
  headers: {
    Accept: 'application/json',
    'Content-Type': 'application/json',
  },
});

export const { CancelToken } = axios;

export const CUSTOMERS_PATH = '/customers/:id';
export const SUPPLIERS_PATH = '/suppliers/:id';

export default client;

export interface Customer {
  id: number,
  name?: string,
  representative?: string,
  kvk_nr?: string,
  email?: string,
  phone?: string,
  phone2?: string,
  street?: string,
  street_extra?: string,
  city?: string,
  country?: string,
  state?: string,
  zip?: string,
  street2?: string,
  street_extra2?: string,
  city2?: string,
  country2?: string,
  state2?: string,
  zip2?: string,
  is_partner?: number,
}

export interface Supplier {
  id: number,
  name?: string,
  representative?: string,
  email?: string,
  phone?: string,
  phone2?: string,
  street?: string,
  street_extra?: string,
  city?: string,
  country?: string,
  state?: string,
  zip?: string,
  street2?: string,
  street_extra2?: string,
  city2?: string,
  country2?: string,
  state2?: string,
  zip2?: string,
  partner?: string,
}
