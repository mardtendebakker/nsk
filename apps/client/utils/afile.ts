import { AFile } from './axios/models/aFile';

export function buildAFileLink(afile: AFile): string {
  return `https://${afile.unique_server_filename}.s3.amazonaws.com/${afile.discr}/${afile.original_client_filename}`;
}
