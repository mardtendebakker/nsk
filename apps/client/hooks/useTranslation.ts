import { useEffect, useState } from 'react';
import { trans, setConfig } from 'itranslator';
import { setDefaultOptions } from 'date-fns';
import {
  nl, enUS, de, fr, ar, faIR,
} from 'date-fns/locale';
import Config from 'itranslator/lib/type/config';
import { getDefaultLocale, setDefaultLocale } from '../utils/storage';
import EventEmitter from '../utils/eventEmitter';
import nlSource from '../public/translations/nl';
import enSource from '../public/translations/en';
import deSource from '../public/translations/de';
import frSource from '../public/translations/fr';
import arSource from '../public/translations/ar';
import faSource from '../public/translations/fa';

export const LOCALE_UPDATED = 'LOCALE_UPDATED';
export type Trans = (key: string, config?: Config | undefined) => string;

export const localeMapping = {
  nl: nlSource,
  en: enSource,
  de: deSource,
  ar: arSource,
  fr: frSource,
  fa: faSource,
};

export const dateFnsMapping = {
  nl,
  en: enUS,
  de,
  fr,
  ar,
  fa: faIR,
};

class LocaleEmitter extends EventEmitter {
  #locale = getDefaultLocale() || 'nl';

  constructor() {
    super();
    setDefaultOptions({ locale: dateFnsMapping[this.#locale] });
  }

  get locale() { return this.#locale; }

  emit(event: string, payload: string) {
    if (LOCALE_UPDATED === event) {
      this.#locale = payload;

      setConfig({ source: localeMapping[payload] });
      setDefaultOptions({ locale: dateFnsMapping[payload] });
      setDefaultLocale(payload);
    }

    return super.emit(event, payload);
  }
}

export const localeStore = new LocaleEmitter();

const useTranslation = (): { updateLocale: (locale: string)=> void, trans: Trans, locale: string } => {
  const [locale, setLocale] = useState<string>(localeStore.locale);

  useEffect(() => {
    localeStore.on(LOCALE_UPDATED, setLocale);
    return () => {
      localeStore.removeListener(LOCALE_UPDATED, setLocale);
    };
  }, [setLocale]);

  const updateLocale = (newLocale: string) => {
    localeStore.emit(LOCALE_UPDATED, newLocale);
    setLocale(newLocale);
  };

  return {
    updateLocale,
    trans,
    locale,
  };
};

export default useTranslation;
