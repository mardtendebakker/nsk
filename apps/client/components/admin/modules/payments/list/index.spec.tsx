import {
  render, screen, fireEvent, waitFor,
} from '@testing-library/react';
import { act } from 'react-dom/test-utils';
import useAxios from '../../../../../hooks/useAxios';
import useTranslation from '../../../../../hooks/useTranslation';
import useCart from '../../../../../hooks/useCart';
import { getQueryParam } from '../../../../../utils/location';
import Payments from '.';

// Mock dependencies
jest.mock('next/router', () => ({
  useRouter: () => ({
    push: jest.fn(),
    query: {},
  }),
}));

jest.mock('../../../../../hooks/useAxios');
jest.mock('../../../../../hooks/useTranslation');
jest.mock('../../../../../hooks/useCart');
jest.mock('../../../../../utils/pushURLParams');
jest.mock('../../../../../utils/location');
jest.mock('./list', () => ({
  __esModule: true,
  default: ({
    payments, onPageChange, onRowsPerPageChange, onSubscribe, disabled,
  }) => (
    <div data-testid="mock-list">
      {payments?.map((payment) => (
        <div key={payment.id} data-testid={`payment-${payment.id}`} />
      ))}
      <button onClick={() => onPageChange(2)} data-testid="change-page" type="button" disabled={disabled}>Change Page</button>
      <button onClick={() => onRowsPerPageChange(20)} data-testid="change-rows" type="button" disabled={disabled}>Change Rows</button>
      <button onClick={() => onSubscribe({ id: 1 }, true)} data-testid="subscribe" type="button" disabled={disabled}>Subscribe</button>
    </div>
  ),
}));

describe('Payments Component', () => {
  const mockCall = jest.fn().mockResolvedValue({ data: [], count: 0 });
  const mockCallSubscribe = jest.fn().mockResolvedValue({});
  const mockCallUnsubscribe = jest.fn().mockResolvedValue({});

  beforeEach(() => {
    jest.clearAllMocks();
    (useAxios as jest.Mock).mockImplementation((method) => {
      if (method === 'post') {
        return { call: mockCallSubscribe, performing: false };
      }
      if (method === 'delete') {
        return { call: mockCallUnsubscribe, performing: false };
      }
      return {
        call: mockCall,
        performing: false,
        data: { data: [], count: 0 },
      };
    });

    (getQueryParam as jest.Mock).mockImplementation((param, defaultValue) => defaultValue);
    (useTranslation as jest.Mock).mockImplementation(() => ({
      trans: (key: string) => key,
    }));
    (useCart as jest.Mock).mockImplementation(() => ({
      clearCart: jest.fn(),
    }));
  });

  it('matches snapshot', () => {
    const { asFragment } = render(<Payments />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('initializes with correct default values', () => {
    render(<Payments />);
    expect(mockCall).toHaveBeenCalledWith({
      params: {
        take: 10,
        skip: 0,
      },
    });
  });

  it('handles page change correctly', async () => {
    render(<Payments />);

    await act(async () => {
      fireEvent.click(screen.getByTestId('change-page'));
    });

    expect(mockCall).toHaveBeenCalledWith({
      params: {
        take: 10,
        skip: 10,
      },
    });
  });

  it('handles rows per page change correctly', async () => {
    render(<Payments />);

    await act(async () => {
      fireEvent.click(screen.getByTestId('change-rows'));
    });

    expect(mockCall).toHaveBeenCalledWith({
      params: {
        take: 20,
        skip: 0,
      },
    });
  });

  it('handles subscription correctly', async () => {
    render(<Payments />);

    await act(async () => {
      fireEvent.click(screen.getByTestId('subscribe'));
    });

    expect(mockCallSubscribe).toHaveBeenCalledWith({
      path: expect.stringContaining('1'),
    });

    await waitFor(() => {
      expect(mockCall).toHaveBeenCalled();
    });
  });

  it('clears cart when clearCart query param is 1', () => {
    (getQueryParam as jest.Mock).mockImplementation((param, defaultValue) => {
      if (param === 'clearCart') return '1';
      return defaultValue;
    });

    const mockClearCart = jest.fn();
    (useCart as jest.Mock).mockImplementation(() => ({
      clearCart: mockClearCart,
    }));

    render(<Payments />);
    expect(mockClearCart).toHaveBeenCalled();
  });

  it('displays payments data correctly', () => {
    const mockPayments = [
      { id: 1, amount: 100 },
      { id: 2, amount: 200 },
    ];

    (useAxios as jest.Mock).mockImplementation(() => ({
      call: mockCall,
      performing: false,
      data: { data: mockPayments, count: 2 },
    }));

    render(<Payments />);
    expect(screen.getByTestId('payment-1')).toBeInTheDocument();
    expect(screen.getByTestId('payment-2')).toBeInTheDocument();
  });

  it('handles loading state correctly', () => {
    (useAxios as jest.Mock).mockImplementation(() => ({
      call: mockCall,
      performing: true,
      data: { data: [], count: 0 },
    }));

    render(<Payments />);
    waitFor(() => {
      expect(screen.getByTestId('change-page')).toHaveAttribute('disabled', 'true');
      expect(screen.getByTestId('change-rows')).toHaveAttribute('disabled', 'true');
      expect(screen.getByTestId('subscribe')).toHaveAttribute('disabled', 'true');
    });
  });
});
