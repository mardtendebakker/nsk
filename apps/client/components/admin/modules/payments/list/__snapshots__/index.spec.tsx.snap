// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Payments Component matches snapshot 1`] = `
<DocumentFragment>
  <h4
    class="MuiTypography-root MuiTypography-h4 css-4il6ew-MuiTypography-root"
  >
    payments
  </h4>
  <div
    data-testid="mock-list"
  >
    <button
      data-testid="change-page"
      type="button"
    >
      Change Page
    </button>
    <button
      data-testid="change-rows"
      type="button"
    >
      Change Rows
    </button>
    <button
      data-testid="subscribe"
      type="button"
    >
      Subscribe
    </button>
  </div>
</DocumentFragment>
`;
